numpy/lib/index_tricks.pyi

```diff
@@ -0,0 +1,179 @@
+    order: _OrderCF = ...
+) -> _ArrayOrScalar[intp]: ...
+
+@overload
+def ix_(*args: _NestedSequence[_SupportsDType[_DType]]) -> Tuple[ndarray[Any, _DType], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[str]) -> Tuple[_ArrayND[str_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[bytes]) -> Tuple[_ArrayND[bytes_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[bool]) -> Tuple[_ArrayND[bool_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[int]) -> Tuple[_ArrayND[int_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[float]) -> Tuple[_ArrayND[float_], ...]: ...
+@overload
+def ix_(*args: _NestedSequence[complex]) -> Tuple[_ArrayND[complex_], ...]: ...
+@overload
+def ix_(*args: _RecursiveSequence) -> Tuple[_ArrayND[Any], ...]: ...
+
+class nd_grid(Generic[_BoolType]):
+    sparse: _BoolType
+    def __init__(self, sparse: _BoolType = ...) -> None: ...
+    @overload
+    def __getitem__(
+        self: nd_grid[Literal[False]],
+        key: Union[slice, Sequence[slice]],
+    ) -> _ArrayND[Any]: ...
+    @overload
+    def __getitem__(
+        self: nd_grid[Literal[True]],
+        key: Union[slice, Sequence[slice]],
+    ) -> List[_ArrayND[Any]]: ...
+
+class MGridClass(nd_grid[Literal[False]]):
+    def __init__(self) -> None: ...
+
+mgrid: MGridClass
+
+class OGridClass(nd_grid[Literal[True]]):
+    def __init__(self) -> None: ...
+
+ogrid: OGridClass
+
+class AxisConcatenator:
+    axis: int
+    matrix: bool
+    ndmin: int
+    trans1d: int
+    def __init__(
+        self,
+        axis: int = ...,
+        matrix: bool = ...,
+        ndmin: int = ...,
+        trans1d: int = ...,
+    ) -> None: ...
+    @staticmethod
+    @overload
+    def concatenate(  # type: ignore[misc]
+        *a: ArrayLike, axis: SupportsIndex = ..., out: None = ...
+    ) -> _ArrayND[Any]: ...
+    @staticmethod
+    @overload
+    def concatenate(
+        *a: ArrayLike, axis: SupportsIndex = ..., out: _ArrayType = ...
+    ) -> _ArrayType: ...
+    @staticmethod
+    def makemat(
+        data: ArrayLike, dtype: DTypeLike = ..., copy: bool = ...
+    ) -> _Matrix: ...
+
+    # TODO: Sort out this `__getitem__` method
+    def __getitem__(self, key: Any) -> Any: ...
+
+class RClass(AxisConcatenator):
+    axis: Literal[0]
+    matrix: Literal[False]
+    ndmin: Literal[1]
+    trans1d: Literal[-1]
+    def __init__(self) -> None: ...
+
+r_: RClass
+
+class CClass(AxisConcatenator):
+    axis: Literal[-1]
+    matrix: Literal[False]
+    ndmin: Literal[2]
+    trans1d: Literal[0]
+    def __init__(self) -> None: ...
+
+c_: CClass
+
+class IndexExpression(Generic[_BoolType]):
+    maketuple: _BoolType
+    def __init__(self, maketuple: _BoolType) -> None: ...
+    @overload
+    def __getitem__(  # type: ignore[misc]
+        self: IndexExpression[Literal[True]], item: slice
+    ) -> Tuple[slice]: ...
+    @overload
+    def __getitem__(self, item: _SliceOrTuple) -> _SliceOrTuple: ...
```

Comment:
This is arguably a little strict, considering `s_` and `index_exp` can, in practice, take any object. The annotations are in line with the objects' intent though.

Please let me know if I should remove the `slice` constraint.

Reply:
I like being strict in cases like these. It helps users write better code. If there's really another type of object that's common and valid (I don't think there is in this case), we can always consider adding it later.

Reply:
This is too strict, as it doesn't allow any of:
* `np.s_[1]`
* `np.s_[...]`
* `np.s_[True]`
* `np.s_[np.int64(1)]`

Reply:
> This is too strict, as it doesn't allow any of these reasonable expressions

That's a fair point. Expanding the type-restrictions will be fairly cumbersome;
removing them in their entirety on the other hand should be trivial.

Reply:
So as of https://github.com/numpy/numpy/pull/18397/commits/01f88a75a5415a40c6e9c386ccb27a91d4f39344 the constraints have been lifted. 

All in all I'd say this is not too bad of a change, as the relevant method can, in fact, take any object 
and its signature can be described exactly.